Deploy SqueezeNet 1.0 INT8 (opset 12) model with ONNX Runtime on Azure Cobalt 100 (#6)
odidev/arm-learning-paths: content/learning-paths/servers-and-cloud-computing/onnx-on-azure/benchmarking.md

weight: 5

### FIXED, DO NOT MODIFY
layout: learningpathall
---

## Test the ONNX based application for Performance
Now that you’ve set up and run the ONNX model (e.g., SqueezeNet), you can use it to benchmark inference performance using Python-based timing or tools like onnxruntime_perf_test. This helps evaluate the ONNX Runtime efficiency on Azure Arm64-based Cobalt 100 instances.
You can also compare the inference time between Cobalt 100 (Arm64) and similar D-series x86_64-based VMs on Azure.
As noted before, the steps to benchmark remain the same, whether it's a Docker container or a custom VM.

## Run the performance tests using onnxruntime_perf_test
The onnxruntime_perf_test is a performance benchmarking tool included in the ONNX Runtime source code. It is used to measure the inference performance of ONNX models under various runtime conditions (like CPU, GPU, or other execution providers). 

### Install Required Build Tools

```console
$ tdnf install -y cmake make gcc-c++ git
```
#### Install Protobuf

```console
$ tdnf install -y protobuf protobuf-devel
```
Then verify:
```console
$ protoc --version
```
You should see something like:

```output
libprotoc x.x
```

> If installation via the package manager fails, or the version is too old for ONNX Runtime; then proceed with [installing Protobuf using the Aarch64 pre-built zip artifact](https://github.com/protocolbuffers/protobuf/releases), as discussed below.

Reviewer: remove the from "via the package manager"

---
title: Benchmarking via onnxruntime_perf_test
weight: 5

### FIXED, DO NOT MODIFY
layout: learningpathall
---

## Test the Onnx based application for Performance
Now that you’ve set up and run the ONNX model (e.g., SqueezeNet), you can use it to benchmark inference performance using Python-based timing or tools like onnxruntime_perf_test. This helps evaluate the ONNX Runtime efficiency on Azure Arm64-based Cobalt 100 instances.
You can also compare the inference time between Cobalt 100 (Arm64) and similar D-series x86_64-based VMs on Azure.
As noted before, the steps to benchmark remain the same, whether it's a Docker container or a custom VM.

## Run the performance tests using onnxruntime_perf_test
The onnxruntime_perf_test is a performance benchmarking tool included in the ONNX Runtime source code. It is used to measure the inference performance of ONNX models under various runtime conditions (like CPU, GPU, or other execution providers). 

### Install Required Build Tools

```console
$ tdnf install -y cmake make gcc-c++ git
```

> ### Try Installing protobuf from Package

Author: ```suggestion
### Install Protobuf
```

> ### Try Installing protobuf from Package

Author: It's a sub-heading under "Install Required Build Tools"

```console
$ tdnf install -y protobuf protobuf-devel
```
Then verify:
```console
$ protoc --version
```
You should see something like:

```output
libprotoc 3.X.X
```

> If the command fails or the version is too old for ONNX Runtime, proceed with building from source.

Author: ```suggestion
If installation via the package manager fails, or the version is too old for ONNX Runtime; then proceed with building from source, as discussed below.
```

> ###  Build Protobuf from Source (Safe on Azure Linux)

Author: ```suggestion
###  Build Protobuf from the source code
```

```console
$ git clone --branch v21.12 https://github.com/protocolbuffers/protobuf.git
$ cd protobuf
$ git submodule update --init --recursive
$ mkdir build && cd build
$ cmake .. -DCMAKE_BUILD_TYPE=Release
$ make -j$(nproc)
$ sudo make install
$ sudo ldconfig
```
Then verify:
```console
$ protoc --version
```

output:
```output
libprotoc 25.3
```

> You should now see the correct version (e.g., libprotoc 25.).

Author: remove it

### Clone and Build ONNX Runtime from Source:

> 

Author: **Remove this note, and add below as intro for this heading:**

The benchmarking tool, **onnxruntime_perf_test**, isn’t available as a pre-built binary artifact for any platform. So, you have to build it from the source, which is expected to take around 40-50 minutes. 
Follow the below to build onnxruntime from the source code, and hence benchmark the module.

> {{% notice Note %}}

Author: remove it.

onnxruntime_perf_test isn’t available as a pre-built binary artifact for any platform. So, you have to build it from the source, which is expected to take around 40-50 minutes. 
{{% /notice %}}

```console
$ tdnf install -y protobuf-compiler libprotobuf-dev libprotoc-dev 
$ git clone --recursive https://github.com/microsoft/onnxruntime 
$ cd onnxruntime
```
Now, build the benchmark by using below command:

```console
./build.sh --config Release --build_dir build/Linux --build_shared_lib --parallel --build --update --skip_tests 
```

> output:

Author: ```suggestion
Output:
```

```output
[100%] Building CXX object CMakeFiles/onnxruntime_test_all.dir/home/azureuser/new/onnxruntime/onnxruntime/test/lora/lora_test.
[100%] Building CXX object CMakeFiles/onnxruntime_test_all.dir/home/azureuser/new/onnxruntime/onnxruntime/test/unittest_main/tain.cc.o
[100%] Linking CXX executable onnxruntime_test_all
[100%] Built target onnxruntime_test_all
2025-07-09 07:09:04,831 build [INFO] - Build complete
```

> This confirms that the onnx benchmarking is equally available to be used on Linux based platforms. 

Author: remove this line too.

### Run the benchmark

> To run ONNX benchmarking on Azure virtual machines or containers, use the same scripts and methodology across both CPU architectures:

Author: Now that the benchmarking tool has been built, you can benchmark the **squeezenet-int8.onnx** model, as below:

> To run ONNX benchmarking on Azure virtual machines or containers, use the same scripts and methodology across both CPU architectures:

Author: ```suggestion
Now that the benchmarking tool has been built, you can benchmark the **squeezenet-int8.onnx** model, as below:
```

```console
$ ./build/Linux/Release/onnxruntime_perf_test -e cpu -r 100 -m times -s -Z -I /home/ubuntu/lerning-path/squeezenet-int8.onnx
```

> #### Command Flags Explanation:

Author: remove this sub-heading.

- e cpu: Use the CPU execution provider (not GPU or any other backend). 
- r 100: Run 100 inferences. 

> - m times: Use "repeat N times" mode. 

Author: ```suggestion
- **m times**: Use "repeat N times" mode. 
```

> - s: Show detailed statistics. 

Author: ```suggestion
- **s**: Show detailed statistics. 
```

> - Z: Disable intra-op thread spinning (reduces CPU usage when idle between runs). 

Author: ```suggestion
- **Z**: Disable intra-op thread spinning (reduces CPU usage when idle between runs). 
```

> - I: Input the ONNX model path without using input/output test data.

Author: ```suggestion
- **I**: Input the ONNX model path without using input/output test data.
```

| **Metric**                | **Value on Docker Container**         | **Value on Virtual Machine**                |
|---------------------------|---------------------------------------|---------------------------------------------|
| **Average Inference Time**| 1.9183 ms                              | 1.9169 ms                                    |
| **Throughput**            | 521.09 inferences/sec                  | 521.41 inferences/sec                        |
| **CPU Utilization**       | 98%                                   | 100%                                         |
| **Peak Memory Usage**     | 35.36 MB                               | 33.57 MB                    |
| **P50 Latency**           | 1.9165 ms                              | 1.9168 ms                                    |
| **Max Latency**           | 2.0142 ms                              | 1.9979 ms                                    |
| **Latency Consistency**   | Consistent                             | Consistent                                   |

> | **Model Used**            | `squeezenet-int8.onnx`                 | `squeezenet-int8.onnx`                       |

Author: remove this "Model Used from the table"
